pkg/cloudproduct/cloudproduct.go

```diff
@@ -58,10 +62,60 @@ const (
 var (
 	logger           = runtime.NewLoggerWithSource("cloudproduct")
 	productDetectors = []func(context.Context, *kubernetes.Clientset) string{gke.Detect}
+
+	// singleton cloud product interface
+	cloudProduct CloudProductControllerHooks
+
+	initializeCloudProductOnce sync.Once
 )
 
-// New instantiates a new CloudProduct interface by product name.
-func New(ctx context.Context, product string, kc *kubernetes.Clientset) (CloudProduct, error) {
+// CloudProductBindFlags binds the --cloud-product flag.
+func CloudProductBindFlags() {
+	viper.SetDefault(cloudProductFlag, autoDetect)
+	pflag.String(cloudProductFlag, viper.GetString(cloudProductFlag), "Cloud product. Set to 'auto' to auto-detect, set to 'generic' to force generic behavior, set to 'gke-autopilot' for GKE Autopilot. Can also use CLOUD_PRODUCT env variable.")
+}
+
+// CloudProductBindEnv binds the CLOUD_PRODUCT env variable.
+func CloudProductBindEnv() error {
+	return viper.BindEnv(cloudProductFlag)
+}
+
+// InitializeCloudProduct initializes the cloud product from the viper flag.
+// Caller must use CloudProductBindFlags and CloudProductBindEnv prior.
+// If this function returns an error, caller should exist, as calling CloudProduct()
```

Comment:
```suggestion
// If this function returns an error, caller should exit, as calling CloudProduct()
```
?

Reply:
Uh, yeah. I mean the caller should exist, too, but I did not mean to make this philosophical. :)